src/Ouroboros-Mini_Protocols-Chain_Sync.thy

```diff
@@ -0,0 +1,226 @@
+
+subsection \<open>Points\<close>
+
+type_synonym point = hash
+
+definition block_point :: "block \<Rightarrow> point" where
+  [simp]: "block_point \<equiv> hash"
+
+subsection \<open>Protocol Parameters\<close>
+
+datatype phase =
+  is_intersection_finding: IntersectionFinding |
+  is_chain_update: ChainUpdate
+
+type_synonym consumer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  phase \<comment> \<open>current phase\<close> \<times>
+  (chain \<Rightarrow> point list)" \<comment> \<open>function for selecting candidate points\<close>
+
+type_synonym producer_parameters = "
+  chain \<comment> \<open>current chain\<close> \<times>
+  chain_index \<comment> \<open>read pointer\<close> \<times>
+  bool \<comment> \<open>whether the next response to \<open>RequestNext\<close> should be \<open>RollBackward\<close>\<close>"
+
+locale chain_sync =
+  fixes initial_params\<^sub>c :: consumer_parameters
+    and initial_params\<^sub>p :: producer_parameters
+
+subsection \<open>Parties\<close>
+
+datatype party =
+  Producer |
+  Consumer
+
+subsection \<open>State Machines\<close>
+
+datatype state =
+  Idle |
+  Intersect |
+  CanAwait
+
+datatype message =
+  is_next_request: RequestNext |
+  is_roll_forward: RollForward \<open>block\<close> \<open>block\<close> |
+  is_roll_backward: RollBackward \<open>point\<close> \<open>block\<close> |
+  is_await_reply: AwaitReply |
+  is_find_intersect: FindIntersect \<open>point list\<close> |
+  is_intersect_found: IntersectFound \<open>point\<close> \<open>block\<close> |
+  is_intersect_not_found: IntersectNotFound \<open>block\<close>
+
+fun agent_in_state' where
+  "agent_in_state' Idle = Consumer" |
+  "agent_in_state' _ = Producer"
+
+inductive can_finish_in_state' where
+  "can_finish_in_state' Idle"
+
+declare can_finish_in_state'.simps [simp]
+
+primrec next_state' where
+  "next_state' Idle m = (
+    partial_case m of
+      FindIntersect _ \<Rightarrow> Intersect |
+      RequestNext \<Rightarrow> CanAwait)" |
+  "next_state' Intersect m = (
+    partial_case m of
+      IntersectFound _ _ \<Rightarrow> Idle |
+      IntersectNotFound _ \<Rightarrow> Idle)" |
+  "next_state' CanAwait m = (
+    partial_case m of
+      RollForward _ _ \<Rightarrow> Idle |
+      RollBackward _ _ \<Rightarrow> Idle |
+      AwaitReply \<Rightarrow> Idle)" \<comment> \<open>only for this initial implementation\<close>
+
+definition state_machine where
+  [simp]: "state_machine = \<lparr>
+    initial_state = Idle,
+    agent_in_state = agent_in_state',
+    can_finish_in_state = can_finish_in_state',
+    next_state = next_state'
+  \<rparr>"
+
+sublocale chain_sync \<subseteq> protocol_state_machine \<open>state_machine\<close> .
+
+subsection \<open>Programs\<close>
+
+definition rollback_to :: "chain \<Rightarrow> point \<Rightarrow> chain" where
+  [simp]: "rollback_to \<C> point = takeWhile (\<lambda>block. hash block \<le> point) \<C>"
```

Comment:
If we keep points completely abstract, as we’ve decided to do meanwhile, then this definition cannot work anymore, because we don’t have access to an order on points.

That said, I already consider this definition to be problematic in its own right. The problem I see is that it is also defined for the illegal case that the given point is not covered by the given chain and the other illegal case that the chain contains multiple items with the same point. With dependent types, you would probably exclude these cases by giving the function a more restrictive type. With Isabelle/HOL, the appropriate way is to leave the function unspecified for illegal input. Otherwise, you might be able to accidentally prove statements about illegal situations by relying on implementation details.

How can we make the function unspecified for illegal input? One way is to append an `if` clause to the defining equation, thus making it conditional. This is what I did with the `follow_up` function in the mini-protocol framework. Leaving out the condition there would have resulted in `follow_up` returning `Done` whenever the message argument is `Done`, even if the given possibilities don’t allow a `Done` message to be transmitted next. However, in the case of `rollback_to`, a more elegant solution might be to use a concise, highly declarative specification like `THE 𝒞₁. ∃𝒞₂. 𝒞 = 𝒞₁ @ 𝒞₂ ∧ point (last 𝒞₁) = p`, which is automatically unspecified if there isn’t exactly one item with the point in question in the chain.